### DISCUSSION TITLE
Inverse dynamics module or extension

### DISCUSSION THREAD

**Discussion by felixchenier** [kineticstoolkit/kineticstoolkit#129] [2. Development: Ideas, new features, doc, etc.]
One of the most requested features to date is a way to do inverse dynamics calculation. It could be either via a tutorial or an actual module (or extension). There is nothing particularly complex that would prevent it from happening. I also see nothing that could be considered controversial: as long as we make our own assumptions on the body's inertial characteristics, and that the provided method is powerful enough to accept any inertial matrix (e.g., not only a mass and a single radius of gyration), then the rest is purely Newtonian physics.

The difficult part is to create an easy yet complete API to do it. For example, the `kinematics` module has three very simple functions to reconstruct track and reconstruct markers (`create_cluster`, `expand_cluster`, `track_cluster`), that are powerful yet easy to understand. My aim is to get the same simplicity with inverse dynamics.

I start this discussion as a notepad for ideas (mine and others), before it turns in a proper feature request. Do not hesitate to put ideas and references.

**Comment by Ipuch** [comment]
Hey, I was thinking maybe you could use [biorbd](http://github.com/pyomeca/biorbd) to run inverse_dynamics. 

You would have to convert the ktk model into a .bioMod file (easy code) through printing command and run inverse dynamics based on joint angles. Let me know if you are interested in a more detail explanation. It would avoid rewriting and revalidating the complex algorithmic of RNEA.

Regards,

Pierre

**Reply by felixchenier** [reply]
@Ipuch Yes I think I need to evaluate this.

To be clear, ktk and biordb do not share the same level of complexity/abstraction. KTK is more lowlevel and non-specific (timeseries, rigid transforms, filtering, etc.). A bit like pyomeca but using different constructs (mostly non-OOP). Contrarily to biordb, there is no model in ktk, and even no inherent parent-children relations between segments. All we do is playing with markers expressed in different independent moving frames.

A year ago, I started an inverse dynamics extension for ktk (https://github.com/felixchenier/kineticstoolkit_inversedynamics) that works, based on R. Dumas, R. Aissaoui, and J. A. De Guise, "A 3D generic inverse dynamic method using wrench notation and quaternion algebra,” Comput Meth Biomech Biomed Eng, vol. 7, no. 3, pp. 159–166, 2004. But I'm not yet happy with it, particularly with the interface (API), precisely because ktk does not define a way to represent models and I'm not even sure I want to include modelling right into ktk (in fact I'm sure I don't want to - I must stick to the general-processing idea of ktk).

If using an external tool is the way to go, then I'm all for it. In that view, using ktk in parallel to biordb seems very logical because there seems to be absolutely no overlap between both libraries. From my understanding, we can even use biordb directly with ktk's TimeSeries since ktk's TimeSeries are just bunches of NumPy arrays with some yummy goodies. Perhaps all we need, in reality, is a good tutorial, some real-life sample data, and readily accessible starter models. Such tutorials would have a place of choice on https://kineticstoolkit.uqam.ca.

Maybe a Zoom call would be interesting, to discuss on the form this could take. I think this articulates well with the planned opening of your upper-body MSK OpenSim models, that may (?) be converted to bioMOD models for such tutorials. Would you be interested in a Zoom call?
